atomics/include/desul/atomics/Compare_Exchange_CUDA.hpp

```diff
@@ -160,5 +161,67 @@ __device__ typename std::enable_if<sizeof(T) == 8, T>::type atomic_compare_excha
 }
 }
 #endif
+
+#if defined(__CUDA_ARCH__) || !defined(__NVCC__)
+namespace desul {
+template <typename T, class MemoryOrder, class MemoryScope>
+DESUL_INLINE_FUNCTION 
+__device__ typename std::enable_if<(sizeof(T) != 8) && (sizeof(T) != 4), T>::type atomic_compare_exchange(
+    T* const dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  // This is a way to avoid dead lock in a warp or wave front
+  T return_val;
+  int done = 0;
+  unsigned int mask = DESUL_IMPL_ACTIVEMASK;
+  unsigned int active = DESUL_IMPL_BALLOT_MASK(mask, 1);
+  unsigned int done_active = 0;
+  while (active != done_active) {
+    if (!done) {
+      if (Impl::lock_address_cuda((void*)dest, scope)) {
+        if(std::is_same<MemoryOrder,MemoryOrderSeqCst>::value) atomic_thread_fence(MemoryOrderRelease(),scope);
```

Comment:
Comment why it is ok to pay for that runtime check.

Reply:
which check? That std::is_same is a compile time check in the end.

Reply:
Are you saying that it gets optimized away?

Reply:
It ought to be, the value in the if is a compile time constant.  We could make all of these a call to an overloaded function instead if there's some worry about that though since it can be dispatched by type.

Reply:
Actually, having a pre_fence/post_fence and just setting them up for dispatch might not be a bad thing for keeping the checks centralized.